joblib/numpy_pickle.py

```diff
@@ -1,229 +1,251 @@
-        if isinstance(s, bytes):
-            return s
-        return s.encode('latin1')
-else:
-    Unpickler = pickle.Unpickler
-    Pickler = pickle.Pickler
-    asbytes = str
 
+class NumpyArrayWrapper(object):
+    """An object to be persisted instead of numpy arrays.
 
-def hex_str(an_int):
-    """Converts an int to an hexadecimal string
+    This object is in charge of:
+    * carrying the information of the persisted array:
+    the subclass, shape, order, dtype of the array. Those
+    ndarray metadata are used to correctly reconstruct the array with numpy
+    functions.
+    * determining if memmap is allowed on the array.
+    * reading the array bytes from a file.
+    * reading the array using memorymap from a file.
+    * writing the array bytes to a file.
+
+    Attributes
+    ----------
+    subclass: numpy.ndarray subclass
+        Determine the subclass of the wrapped array.
+    shape: numpy.ndarray shape
+        Determine the shape of the wrapped array.
+    order: str
+        Determine the order of wrapped array data. Allowed values are 'F' for
+        fortran order and 'C' for C order.
+    dtype: numpy.ndarray dtype
+        Determine the data type of the wrapped array.
+    allow_mmap: bool
+        Determine if memory mapping is allowed on the wrapped array.
+        Default: False.
     """
-    return '{0:#x}'.format(an_int)
-
 
-_MEGA = 2 ** 20
+    def __init__(self, subclass, shape, order, dtype, allow_mmap=False):
+        """Constructor. Store the useful information for later."""
+        self.subclass = subclass
+        self.shape = shape
+        self.order = order
+        self.dtype = dtype
+        self.allow_mmap = allow_mmap
 
-# Compressed pickle header format: _ZFILE_PREFIX followed by _MAX_LEN
-# bytes which contains the length of the zlib compressed data as an
-# hexadecimal string. For example: 'ZF0x139              '
-_ZFILE_PREFIX = asbytes('ZF')
-_MAX_LEN = len(hex_str(2 ** 64))
+    def write_array(self, array, pickler):
+        """Write array bytes to pickler file handle.
 
+        This function is an adapation of the numpy write_array function
+        available in version 0.10 in numpy/lib/format.py.
+        We added some code to support versions of numpy prior to 1.9.
+        """
+        # Set buffer size to 16 MiB to hide the Python loop overhead.
+        buffersize = max(16 * 1024 ** 2 // array.itemsize, 1)
+        np_ver = [int(x) for x in pickler.np.__version__.split('.', 2)[:2]]
+
+        if array.dtype.hasobject:
+            # We contain Python objects so we cannot write out the data
+            # directly. Instead, we will pickle it out with version 2 of the
+            # pickle protocol.
+            pickle.dump(array, pickler.file, protocol=2)
+        elif self.order == 'F':
+            if pickler.np.compat.isfileobj(pickler.file):
+                array.T.tofile(pickler.file)
+            else:
+                # 'tobytes' is available since numpy 1.9.
```

Comment:
> Also note that in the case of fortran order, the .tostring() is not called exactly the same way : it uses the transposed version of the array, but still with 'C' order.

AFAIK, .tostring is just a deprecated alias for .tobytes, see http://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.tostring.html for example. So you could use .tostring to be compatible with all the versions of numpy we are supporting.